lib/spack/spack/solver/asp.py

```diff
@@ -2867,6 +2763,170 @@ def _specs_from_requires(self, pkg_name, section):
                 yield _spec_with_default_name(s, pkg_name)
 
 
+class RequirementParser:
+    """Parses requirements from package.py files and configuration, and returns rules."""
+
+    def __init__(self, configuration):
+        self.config = configuration
+
+    def rules(self, pkg: "spack.package_base.PackageBase") -> List[RequirementRule]:
+        result = []
+        result.extend(self.rules_from_package_py(pkg))
+        result.extend(self.rules_from_require(pkg))
+        result.extend(self.rules_from_prefer(pkg))
+        result.extend(self.rules_from_conflict(pkg))
+        return result
+
+    def rules_from_package_py(self, pkg) -> List[RequirementRule]:
+        rules = []
+        for when_spec, requirement_list in pkg.requirements.items():
+            for requirements, policy, message in requirement_list:
+                rules.append(
+                    RequirementRule(
+                        pkg_name=pkg.name,
+                        policy=policy,
+                        requirements=requirements,
+                        kind=RequirementKind.PACKAGE,
+                        condition=when_spec,
+                        message=message,
+                    )
+                )
+        return rules
+
+    def rules_from_virtual(self, virtual_str: str) -> List[RequirementRule]:
+        requirements = self.config.get("packages", {}).get(virtual_str, {}).get("require", [])
+        return self._rules_from_requirements(
+            virtual_str, requirements, kind=RequirementKind.VIRTUAL
+        )
+
+    def rules_from_require(self, pkg: "spack.package_base.PackageBase") -> List[RequirementRule]:
+        kind, requirements = self._raw_yaml_data(pkg, section="require")
+        return self._rules_from_requirements(pkg.name, requirements, kind=kind)
+
+    def rules_from_prefer(self, pkg: "spack.package_base.PackageBase") -> List[RequirementRule]:
+        result = []
+        kind, preferences = self._raw_yaml_data(pkg, section="prefer")
+        for spec_str in preferences:
+            spec = sc.parse_spec_from_yaml_string(spec_str)
+            result.append(
+                # A strong preference is defined as:
+                #
+                # require:
+                # - any_of: [spec_str, "@:"]
+                RequirementRule(
+                    pkg_name=pkg.name,
+                    policy="any_of",
+                    requirements=[spec, spack.spec.Spec("@:")],
+                    kind=kind,
+                    message=None,
+                    condition=spack.spec.Spec(),
+                )
+            )
+        return result
+
+    def rules_from_conflict(self, pkg: "spack.package_base.PackageBase") -> List[RequirementRule]:
```

Comment:
Does it actually make sense to pursue this weird `one_of(x, "@:")` construct if you have dedicated syntax for it? Is there a downside to emitting an ordinary conflict?

Reply:
On success the two statements should be semantically equivalent: a solution is not valid if a conflict has a match. On failure using one or the other may have an impact on the produced error message, due to the higher priority requests have wrt conflicts. 

Also, I didn't benchmark which of the two formulations is faster - in case there is a clear winner. Do we have preferences on how to proceed here?

Reply:
Maybe good to benchmark if it's not a lot of work, otherwise we'll probably forget about it.

Reply:
I think we should do the benchmark for this in a follow-on -- @alalazo can you do that?